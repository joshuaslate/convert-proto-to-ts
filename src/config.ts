import { findUp } from 'find-up';
import fs from 'fs/promises';
import { cli } from './cli';

export enum ProtoSource {
  Git = 'Git',
  FileSystem = 'File System',
}

export enum TypeNameCase {
  Camel = 'camelCase',
  Constant = 'constantCase',
  Pascal = 'pascalCase',
  Snake = 'snakeCase',
}

export interface Config {
  // tempDir is the path, relative to the root, that the proto files are cloned into if an option is provided for protoGitRepository
  tempDir: string;
  // protoGitRepository is the repository that will be git cloned in order to pull proto definitions
  protoGitRepository?: string;
  // protoPath is the path, relative to the root, to proto definitions on your local file system
  protoPath?: string;
  // outputPath is the path, relative to the root, where the generated TypeScript files will be saved
  outputPath: string;
  // fieldNameKeepCase is true if you'd like to keep field casing instead of camel casing (passed to protobufjs loadSync options), defaults to false
  fieldNameKeepCase?: boolean;
  // namespacesToIgnore is an array of namespaces that you would like to avoid resolving types for, e.g., ["google"]
  namespacesToIgnore?: string[];
  /**
   * typeNameCase determines the casing for generated interface types
   *  Available options:
   *    camelCase (e.g., camelCase)
   *    constantCase (e.g., CONSTANT_CASE)
   *    pascalCase (e.g., PascalCase)
   *    snakeCase (e.g., snake_case)
   *
   *    default: pascalCase
   */
  typeNameCase: TypeNameCase;
  /**
   * enumNameCase determines the casing for generated enum types
   *  Available options:
   *    camelCase (e.g., camelCase)
   *    constantCase (e.g., CONSTANT_CASE)
   *    pascalCase (e.g., PascalCase)
   *    snakeCase (e.g., snake_case)
   *
   *    default: pascalCase
   */
  enumNameCase: TypeNameCase;
  /**
   * typeNameTemplate determines how the generated interface types are named.
   *  Available variables:
   *    parentNodeNames
   *    typeName
   *
   *    default: "{{parentNodeNames}}{{typeName}}"
   */
  typeNameTemplate: string;
  /**
   * enumNameTemplate determines how the generated enum types are named.
   *  Available variables:
   *    parentNodeNames
   *    typeName
   *
   *    default: "{{parentNodeNames}}{{typeName}}"
   */
  enumNameTemplate: string;
  /**
   * typeNameIgnoreParentNodeNames is an array of parent node names in the generated type names that you'd like to leave off.
   * This is only used when {{parentNodeNames}} is included in typeNameTemplate/enumNameTemplate.
   * For example, if the package is your_app.services.auth.v1, and the message name is RegisterRequest, ["your_app", "services"] would yield:
   * export interface AuthV1RegisterRequest {}
   */
  typeNameIgnoreParentNodeNames?: string[];
  // generatedTypeOverrides is a Record<string, string> for type overrides (key == Protobuf type, value == TypeScript type)
  generatedTypeOverrides?: Record<string, string>;
  // generatedTypeComments is a Record<string, string> for leading comments for types, e.g., { 'google.protobuf.Timestamp': 'format: date-time' }
  generatedTypeComments?: Record<string, string>;
  // generateIndexFile is true if an index.ts file exporting all generated types should be generated (default: true)
  generateIndexFile?: boolean;
  /**
   * generateEnumType can be set to either 'union' or 'enum'
   *    union: export type FakeEnum = 'FAKE_ENUM_UNSPECIFIED' | 'FAKE_ENUM_VALUE';
   *    enum: export enum FakeEnum { Unspecified: 'FAKE_ENUM_UNSPECIFIED', Value: 'FAKE_ENUM_VALUE' };
   * Nested enums will always be generated as inline union types, because enums can't be declared inline
   */
  generateEnumType?: 'union' | 'enum';
  /**
   * fileHeaderCommentTemplate can be set to define a header that will be prepended to all generated TypeScript files. Variables available:
   *    generationTimestamp, sourceFile
   * Example: "DO NOT EDIT! Types generated from {{sourceFile}} at {{generationTimestamp}}."
   */
  fileHeaderCommentTemplate?: string;
  /**
   * indexFileHeaderCommentTemplate can be set to define a header that will be prepended to the generated TypeScript file (if enabled). Variables available:
   *    generationTimestamp
   * Example: "DO NOT EDIT! Types generated at {{generationTimestamp}}."
   */
  indexFileHeaderCommentTemplate?: string;
}

export const defaultConfig: Config = {
  tempDir: '__proto-to-ts-temp',
  outputPath: 'generated_from_proto',
  fieldNameKeepCase: false,
  generateIndexFile: true,
  enumNameTemplate: '{{parentNodeNames}}{{typeName}}',
  enumNameCase: TypeNameCase.Pascal,
  typeNameTemplate: '{{parentNodeNames}}{{typeName}}',
  typeNameCase: TypeNameCase.Pascal,
};

function mergeConfigs(cliArgs: Partial<Config>, configFile?: Partial<Config>) {
  const { protoPath, protoGitRepository, ...restOfConfigFile } = configFile || {};

  return {
    ...defaultConfig,
    ...restOfConfigFile,
    protoPath: cliArgs.protoGitRepository ? undefined : protoPath,
    protoGitRepository: cliArgs.protoPath ? undefined : protoGitRepository,
    ...cliArgs,
  };
}

export async function loadConfig(configFromArgs: Partial<Config>): Promise<Config> {
  const foundConfigPath = await findUp('.proto_to_ts_config.json');

  if (foundConfigPath) {
    const fileContent = await fs.readFile(foundConfigPath, 'utf8');

    if (fileContent) {
      try {
        const fileContentAsObject = JSON.parse(fileContent);

        return mergeConfigs(configFromArgs, fileContentAsObject);
      } catch (e) {
        throw new Error(
          `[convert-proto-to-ts]: error encountered while parsing custom .proto_to_ts_config.json file: ${e}`,
        );
      }
    }
  }

  return mergeConfigs(configFromArgs);
}

const CLI_ARGS_TO_CONFIG_KEY: Record<string, keyof Config> = {
  'git-repository': 'protoGitRepository',
  'proto-path': 'protoPath',
};

export function parseCLIArgs(args: string[]): Partial<Config> {
  const config: Partial<Config> = {};
  let configItem: keyof Config | undefined;

  for (let i = 0; i < args.length; i++) {
    const configKey = CLI_ARGS_TO_CONFIG_KEY[args[i].substring(2).toLowerCase()];

    if (configItem) {
      if (!configKey) {
        config[configItem] = args[i] as any;
      }

      configItem = undefined;
    } else {
      if (configKey) {
        configItem = configKey;
      }
    }
  }

  return config;
}
