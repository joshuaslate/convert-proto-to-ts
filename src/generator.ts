import protobuf from 'protobufjs';
// @ts-ignore
import { camelCase, constantCase, pascalCase, snakeCase } from 'change-case';
import path from 'path';
import ts from 'typescript';
import { Config, defaultConfig, TypeNameCase } from './config';

const UNKNOWN_FILE_NAME = 'proto_to_ts_unknown_file_name.proto' as const;

const optionalFieldMarker = ts.factory.createToken(ts.SyntaxKind.QuestionToken);

const UNKNOWN_TYPE = 'any' as const;

interface IndexFileVariables {
  generationTimestamp: string;
}

interface TypeFileVariables {
  generationTimestamp: string;
  sourceFile: string;
}

interface TypeNameVariables {
  parentNodeNames: string;
  typeName: string;
}

function replaceVariables<T = {}>(template: string, variables: T, padVariable?: boolean): string {
  let str = template;

  for (const variableName in variables) {
    let variableValue = (variables as Record<string, string>)[variableName] || '';

    if (padVariable) {
      variableValue = ` ${variableValue} `;
    }
    str = str.replaceAll(new RegExp('{{(?:\\s+)?(' + variableName + ')(?:\\s+)?}}', 'g'), variableValue);
  }

  return str.trim();
}

function getRelativePath(source: string, target: string) {
  const targetArr = target.split('/');
  const sourceArr = source.split('/');
  // Remove filename from end of source & target, discard source
  sourceArr.pop();
  const targetFileName = targetArr.pop();

  const relativePath = path.relative(sourceArr.join('/'), targetArr.join('/'));

  return (relativePath ? `${relativePath}/${targetFileName}` : `./${targetFileName}`).replaceAll(path.sep, '/');
}

function getCaseFunction(caseName: TypeNameCase) {
  return {
    [TypeNameCase.Camel]: camelCase,
    [TypeNameCase.Constant]: constantCase,
    [TypeNameCase.Pascal]: pascalCase,
    [TypeNameCase.Snake]: snakeCase,
  }[caseName];
}

interface CachedType {
  generatedPath: string;
  generatedName: string;
  type: protobuf.Type | protobuf.Enum;
}

type ImportAdder = (namedExport: string, importPath: string) => void;

interface TypeScriptFileToCreate {
  content: string;
  path: string;
}

export class ProtoToTypeScriptGenerator {
  static buildTypeName(
    node: protobuf.Type | protobuf.Enum,
    template: string,
    caseFn: (str: string) => string,
    typeNameIgnoreParentNodeNames?: string[],
  ) {
    const parentNames: string[] = [];

    let currentNode: protobuf.ReflectionObject | null = node.parent;
    while (currentNode) {
      if (!typeNameIgnoreParentNodeNames?.includes(currentNode.name)) {
        parentNames.unshift(currentNode.name);
      }

      currentNode = currentNode.parent;
    }

    return caseFn(
      replaceVariables<TypeNameVariables>(
        template,
        {
          parentNodeNames: parentNames.join(' '),
          typeName: node.name,
        },
        true,
      ),
    );
  }

  static buildTypeCacheKey(node: protobuf.Type | protobuf.Enum) {
    return `${node.filename}:${ProtoToTypeScriptGenerator.buildTypeName(
      node,
      defaultConfig.typeNameTemplate,
      getCaseFunction(defaultConfig.typeNameCase),
    )}`;
  }

  static getEnumKeyName(value: string) {
    // If the first character is a number, wrap in quotes
    if (!Number.isNaN(Number(value[0]))) {
      return `'${value}'`;
    }

    // If the name contains an unsupported character, wrap it in quotes
    for (const char of value) {
      if (!char.match(/\p{Letter}|[0-9]|\$|_/u)) {
        return `'${value}'`;
      }
    }

    return value;
  }

  config: Config;
  root: protobuf.Root;
  typeCache: Record<string, CachedType>;
  wellKnownMappings: Record<string, string>;

  constructor(config: Config, root: protobuf.Root) {
    this.root = root;
    this.config = config;
    this.typeCache = {};

    // As per documentation at https://protobuf.dev/programming-guides/proto3/#json
    this.wellKnownMappings = {
      'bool': 'boolean',
      'double': 'number',
      'float': 'number',
      'int32': 'number',
      'uint32': 'number',
      'sint32': 'number',
      'fixed32': 'number',
      'sfixed32': 'number',
      'string': 'string',
      'int64': 'string',
      'fixed64': 'string',
      'uint64': 'string',
      'sint64': 'string',
      'sfixed64': 'string',
      'bytes': 'string',
      'google.protobuf.Timestamp': 'string',
      'google.protobuf.Duration': 'string',
      'google.protobuf.FieldMask': 'string',
      'google.protobuf.BoolValue': 'boolean | null',
      'google.protobuf.StringValue': 'string | null',
      'google.protobuf.Int64Value': 'string | null',
      'google.protobuf.UInt64Value': 'string | null',
      'google.protobuf.BytesValue': 'string | null',
      'google.protobuf.Int32Value': 'number | null',
      'google.protobuf.UInt32Value': 'number | null',
      'google.protobuf.FloatValue': 'number | null',
      'google.protobuf.DoubleValue': 'number | null',
      'google.protobuf.NullValue': 'null',
      'google.protobuf.Empty': '{}',
      'google.protobuf.ListValue': 'any[]',
      'google.protobuf.Struct': 'Record<string, any>',
      'google.protobuf.Any': "{ '@type': string, value: any }",
      ...(config.generatedTypeOverrides || {}),
    };
  }

  private getFieldTypeMapping(field: protobuf.Field | protobuf.MapField, addImport: ImportAdder): string | undefined {
    const mappedType = this.wellKnownMappings[field.type];

    if (mappedType) {
      return mappedType;
    }

    if (field.resolvedType) {
      const typeResolvedFromCache = this.typeCache[ProtoToTypeScriptGenerator.buildTypeCacheKey(field.resolvedType)];

      if (typeResolvedFromCache) {
        addImport(typeResolvedFromCache.generatedName, typeResolvedFromCache.generatedPath);
        return typeResolvedFromCache.generatedName;
      }
    }
  }

  private convertFieldType(field: protobuf.Field | protobuf.MapField, addImport: ImportAdder): ts.TypeNode | undefined {
    let node: ts.TypeNode;

    // If an enum type was generated (non-nested type), reference it, else create an inline union type declaration
    if (
      field.resolvedType instanceof protobuf.Enum &&
      !this.typeCache[ProtoToTypeScriptGenerator.buildTypeCacheKey(field.resolvedType)]
    ) {
      node = this.generateEnumDefinition(field.resolvedType) as ts.TypeNode;
    } else {
      let mappedType = this.getFieldTypeMapping(field, addImport);

      if (field.map) {
        const mapField = field as protobuf.MapField;
        const keyMappedType = this.getFieldTypeMapping(
          { ...field, type: mapField.keyType } as protobuf.Field,
          addImport,
        );
        mappedType = `Record<${keyMappedType}, ${mappedType}>`;
      }

      if (!mappedType && field.resolvedType instanceof protobuf.Type && field.parent instanceof protobuf.Type) {
        node = this.generateNestedType(field.resolvedType, addImport);
      } else {
        node = ts.factory.createTypeReferenceNode(mappedType || UNKNOWN_TYPE);
      }
    }

    if (field.repeated) {
      node = ts.factory.createArrayTypeNode(node);
    }

    return node;
  }

  private generateEnumDefinition(protoEnum: protobuf.Enum, generatedName?: string) {
    // If generatedName is not provided, a union type will be returned
    if (this.config.generateEnumType === 'enum' && generatedName) {
      return ts.factory.createEnumDeclaration(
        [ts.factory.createModifier(ts.SyntaxKind.ExportKeyword)],
        ts.factory.createIdentifier(generatedName),
        Object.keys(protoEnum.values || {}).map((value) => {
          const destutteredValue = pascalCase(value).replace(pascalCase(protoEnum.name), '');

          return ts.factory.createEnumMember(
            ProtoToTypeScriptGenerator.getEnumKeyName(destutteredValue),
            ts.factory.createStringLiteral(value, true),
          );
        }),
      );
    }

    const literals = Object.keys(protoEnum.values || {}).map((value) => ts.factory.createStringLiteral(value, true));

    const union = ts.factory.createUnionTypeNode(literals as unknown as readonly ts.TypeNode[]);

    if (!generatedName) {
      return union;
    }

    return ts.factory.createTypeAliasDeclaration(
      [ts.factory.createModifier(ts.SyntaxKind.ExportKeyword)],
      ts.factory.createIdentifier(generatedName),
      undefined,
      union,
    );
  }

  private generateFieldMember(
    field: protobuf.Field | protobuf.MapField,
    addImport: ImportAdder,
  ): ts.TypeElement | void {
    const baseFieldType = this.convertFieldType(field, addImport);
    let member = ts.factory.createPropertySignature(
      undefined,
      camelCase(field.name),
      field.optional || !field.required ? optionalFieldMarker : undefined,
      baseFieldType,
    );

    // If the field is part of a oneof, add a comment to indicate the start and end of the oneof,
    // but if proto3_optional is set to true, don't treat it as a oneof (protobuf.js does, probably for backwards compatibility)
    if (field.partOf instanceof protobuf.OneOf && !field.options?.['proto3_optional']) {
      if (field.id === field.partOf.fieldsArray[0].id) {
        member = ts.addSyntheticLeadingComment(
          member,
          ts.SyntaxKind.SingleLineCommentTrivia,
          ` start oneof "${field.partOf.name}"`,
          false,
        );
      }

      if (field.id === field.partOf.fieldsArray[field.partOf.fieldsArray.length - 1].id) {
        member = ts.addSyntheticTrailingComment(
          member,
          ts.SyntaxKind.SingleLineCommentTrivia,
          ` end oneof "${field.partOf.name}"`,
          false,
        );
      }
    }

    return member;
  }

  private generateNestedType(node: protobuf.Type, addImport: ImportAdder) {
    const members: ts.TypeElement[] = [];

    for (const field of node.fieldsArray) {
      const member = this.generateFieldMember(field, addImport);

      if (member) {
        members.push(member);
      }
    }

    return ts.factory.createTypeLiteralNode(members);
  }

  private generateInterfaceMembersFromProtobufType(node: protobuf.Type, addImport: ImportAdder): ts.TypeElement[] {
    const members: ts.TypeElement[] = [];

    if (!node.fieldsArray.length) {
      return members;
    }

    for (const field of node.fieldsArray) {
      let member = this.generateFieldMember(field, addImport);

      if (member) {
        if (this.config.generatedTypeComments?.[field.type]) {
          member = ts.addSyntheticLeadingComment(
            member,
            ts.SyntaxKind.SingleLineCommentTrivia,
            ` ${this.config.generatedTypeComments[field.type]}`,
            false,
          );
        }

        members.push(member);
      }
    }

    return members;
  }

  private generateInterfaceFromProtobufType(
    node: protobuf.Type,
    generatedName: string,
    addImport: ImportAdder,
  ): ts.Node {
    return ts.factory.createInterfaceDeclaration(
      [ts.factory.createModifier(ts.SyntaxKind.ExportKeyword)],
      ts.factory.createIdentifier(generatedName),
      [],
      undefined,
      this.generateInterfaceMembersFromProtobufType(node, addImport),
    );
  }

  private buildGeneratedTypeName(node: protobuf.Type | protobuf.Enum) {
    const caseFn = getCaseFunction(node instanceof protobuf.Enum ? this.config.enumNameCase : this.config.typeNameCase);
    const template = node instanceof protobuf.Enum ? this.config.enumNameTemplate : this.config.typeNameTemplate;

    return ProtoToTypeScriptGenerator.buildTypeName(node, template, caseFn, this.config.typeNameIgnoreParentNodeNames);
  }

  private buildGeneratedFilename(node: protobuf.Type | protobuf.Enum) {
    const pathRoot = this.config.tempDir.replaceAll(path.sep, '/');
    const base = (node.filename || node.parent?.filename || UNKNOWN_FILE_NAME)
      .replace(pathRoot, '')
      .replace('.proto', '.ts');

    return base.startsWith('/') ? base.substring(1) : base;
  }

  private collectTypes(node: protobuf.ReflectionObject) {
    if (!node) {
      return;
    }

    if (node instanceof protobuf.Root) {
      for (const nestedItemName in node.nested) {
        this.collectTypes(node.nested[nestedItemName]);
      }

      return;
    }

    if (node instanceof protobuf.Namespace) {
      if (!this.config.namespacesToIgnore?.includes(node.name)) {
        for (const nestedItemName in node.nested) {
          this.collectTypes(node.nested[nestedItemName]);
        }
      }
    }

    if (node instanceof protobuf.Service) {
      for (const nestedItemName in node.nested) {
        this.collectTypes(node.nested[nestedItemName]);
      }
    }

    if ((node instanceof protobuf.Type || node instanceof protobuf.Enum) && !(node.parent instanceof protobuf.Type)) {
      this.typeCache[ProtoToTypeScriptGenerator.buildTypeCacheKey(node)] = {
        generatedName: this.buildGeneratedTypeName(node),
        generatedPath: this.buildGeneratedFilename(node),
        type: node,
      };
    }
  }

  private getTypesByFile() {
    return Object.values(this.typeCache).reduce(
      (accum, curr) => {
        accum[curr.generatedPath] = [...(accum[curr.generatedPath] || []), curr];
        return accum;
      },
      {} as Record<string, CachedType[]>,
    );
  }

  generateTypeDefinitionsByFile(): TypeScriptFileToCreate[] {
    const generationTimestamp = new Date().toISOString();

    // Collect all types
    this.collectTypes(this.root);

    const files: TypeScriptFileToCreate[] = [];
    const printer = ts.createPrinter({ newLine: ts.NewLineKind.LineFeed });
    const typesByFile = this.getTypesByFile();

    for (const fileToCreate in typesByFile) {
      const imports: Map<string, Set<string>> = new Map();
      const nodeList: ts.Node[] = [];

      function addImport(namedExport: string, importPath: string) {
        if (importPath !== fileToCreate) {
          if (!imports.has(importPath)) {
            imports.set(importPath, new Set());
          }

          imports.get(importPath)!.add(namedExport);
        }
      }

      for (const cachedType of typesByFile[fileToCreate]) {
        if (!(cachedType.type.parent instanceof protobuf.Type)) {
          if (cachedType.type instanceof protobuf.Enum) {
            nodeList.push(
              this.generateEnumDefinition(cachedType.type, cachedType.generatedName),
              ts.factory.createIdentifier('\n'),
            );
          } else if (cachedType.type instanceof protobuf.Type) {
            nodeList.push(
              this.generateInterfaceFromProtobufType(cachedType.type, cachedType.generatedName, addImport),
              ts.factory.createIdentifier('\n'),
            );
          }
        }
      }

      for (const [importFile, namedImports] of imports) {
        nodeList.unshift(
          ts.factory.createImportDeclaration(
            undefined,
            ts.factory.createImportClause(
              true,
              ts.factory.createIdentifier(`{ ${Array.from(namedImports).join(', ')} }`),
              undefined,
            ),
            ts.factory.createStringLiteral(getRelativePath(fileToCreate, importFile).replace('.ts', ''), true),
          ),
        );
      }

      if (imports.size) {
        nodeList.splice(imports.size, 0, ts.factory.createIdentifier('\n'));
      }

      if (this.config.fileHeaderCommentTemplate) {
        nodeList.unshift(
          ts.factory.createJSDocComment(
            replaceVariables<TypeFileVariables>(this.config.fileHeaderCommentTemplate, {
              generationTimestamp,
              sourceFile: fileToCreate.replace('.ts', '.proto'),
            }),
          ),
          ts.factory.createIdentifier('\n'),
        );
      }

      files.push({
        path: path.join(this.config.outputPath, fileToCreate).replaceAll(path.sep, '/'),
        content: printer.printList(
          ts.ListFormat.MultiLine,
          ts.factory.createNodeArray(nodeList),
          ts.createSourceFile(fileToCreate, '', ts.ScriptTarget.ESNext, true, ts.ScriptKind.TS),
        ),
      });
    }

    if (this.config.generateIndexFile) {
      const baseNodes: ts.Node[] = [];

      if (this.config.indexFileHeaderCommentTemplate) {
        baseNodes.push(
          ts.factory.createJSDocComment(
            replaceVariables<IndexFileVariables>(this.config.indexFileHeaderCommentTemplate, { generationTimestamp }),
          ),
        );
      }

      const indexFile = printer.printList(
        ts.ListFormat.MultiLine,
        ts.factory.createNodeArray([
          ...baseNodes,
          ...Object.keys(typesByFile).map((fileCreated) =>
            ts.factory.createExportDeclaration(
              undefined,
              false,
              undefined,
              ts.factory.createStringLiteral(`./${fileCreated.replace('.ts', '')}`, true),
            ),
          ),
        ]),
        ts.createSourceFile('index.ts', '', ts.ScriptTarget.ESNext, true, ts.ScriptKind.TS),
      );

      files.push({
        content: indexFile,
        path: path.join(this.config.outputPath, 'index.ts').replaceAll(path.sep, '/'),
      });
    }

    return files;
  }
}
