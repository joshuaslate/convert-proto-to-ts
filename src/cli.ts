import fs from 'fs';
import path from 'path';
import { loadConfig, parseCLIArgs } from './config';
import { loadAndParseProtos } from './load';
import { ProtoToTypeScriptGenerator } from './generator';

interface Args {
  cwd: string;
  args: string[];
}

export async function cli({ cwd, args }: Args) {
  try {
    const config = await loadConfig(parseCLIArgs(args));
    const root = await loadAndParseProtos(cwd, config);
    const generator = new ProtoToTypeScriptGenerator(config, root);

    const typeDefinitionsByFile = generator.generateTypeDefinitionsByFile();

    if (config.clearOutputPath) {
      console.log('Clearing output path before writing new generated files');
      fs.rmSync(path.join(cwd, config.outputPath), { recursive: true, force: true });
    }

    for (const tsFile of typeDefinitionsByFile) {
      fs.mkdirSync(path.dirname(path.join(cwd, tsFile.path)), { recursive: true });
      fs.writeFileSync(path.join(cwd, tsFile.path), tsFile.content);

      console.log(`Successfully wrote ${tsFile.path}`);
    }
  } catch (e) {
    console.error(e);
  }
}
