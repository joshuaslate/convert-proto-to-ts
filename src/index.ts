import { cli } from './cli';

cli({ cwd: process.cwd() });
