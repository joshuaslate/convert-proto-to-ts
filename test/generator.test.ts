import { fileURLToPath } from 'url';
import { dirname } from 'path';
import { describe, it, expect } from 'vitest';
import { ProtoToTypeScriptGenerator } from '../src/generator';
import { loadAndParseProtos } from '../src/load';
import { loadConfig } from '../src/config';

const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);

describe(ProtoToTypeScriptGenerator, () => {
  it('should generate types from a given directory of protos', async () => {
    const config = await loadConfig();
    config.protoPath = 'protos';
    config.fileHeaderCommentTemplate = undefined;
    config.indexFileHeaderCommentTemplate = undefined;

    const root = await loadAndParseProtos(__dirname, config);
    const generator = new ProtoToTypeScriptGenerator(config, root);

    expect(generator.generateTypeDefinitionsByFile()).toMatchSnapshot();
  });
});
