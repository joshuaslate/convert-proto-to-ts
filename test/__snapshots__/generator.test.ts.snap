// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ProtoToTypeScriptGenerator > should generate types from a given directory of protos 1`] = `
[
  {
    "content": "export interface AuthV1Team {
    id?: string;
    name?: string;
    birthDate?: string;
    // format: date-time
    createdAt?: string;
}

",
    "path": "generated_from_proto/company_name/services/auth/v1/team.ts",
  },
  {
    "content": "import type { AuthV1Team } from './team';
import type { ListifyV1PageRequest, ListifyV1Search, ListifyV1Sort, ListifyV1Filter, ListifyV1PageResponse } from '../../../../listify/v1/fields';

export interface AuthV1CreateUserRequest {
    email?: string;
    profileData?: { '@type': string, value: any };
    mfaSetup?: {
        mfaSecret?: string;
        otp?: string;
    };
}

export interface AuthV1ListUsersRequest {
    teamId?: string;
    page?: ListifyV1PageRequest;
    search?: ListifyV1Search;
    sorts?: ListifyV1Sort[];
    filters?: ListifyV1Filter[];
}

export interface AuthV1ListUsersResponse {
    users?: AuthV1User[];
    page?: ListifyV1PageResponse;
}

export interface AuthV1User {
    id?: string;
    email?: string;
    // format: date-time
    createdAt?: string;
    // format: date-time
    updatedAt?: string;
    metadata?: Record<string, string>;
    isAdmin?: boolean;
    teams?: AuthV1Team[];
    role?: AuthV1Role;
    status?: 'STATUS_UNSPECIFIED' | 'STATUS_CLOSED' | 'STATUS_ACTIVE';
    profile?: {
        language?: string;
        timezone?: string;
        theme?: 'THEME_UNSPECIFIED' | 'THEME_DARK' | 'THEME_LIGHT';
        address?: AuthV1Address;
    };
    // start oneof "permission"
    fullAccess?: boolean;
    scope?: string; // end oneof "permission"
}

export enum AuthV1Role {
    Unspecified = 'ROLE_UNSPECIFIED',
    Customer = 'ROLE_CUSTOMER',
    Admin = 'ROLE_ADMIN'
}

export interface AuthV1Address {
    line1?: string;
    line2?: string;
    city?: string;
    state?: string;
    zip?: string;
    country?: string;
}

",
    "path": "generated_from_proto/company_name/services/auth/v1/user.ts",
  },
  {
    "content": "export interface ListifyV1MethodOptions {
    enable?: boolean;
}

export interface ListifyV1OneofRulesOptions {
    filterable?: boolean;
}

export interface ListifyV1FieldRulesOptions {
    // start oneof "type"
    double?: ListifyV1NumericRules;
    fixed32?: ListifyV1NumericRules;
    fixed64?: ListifyV1NumericRules;
    float?: ListifyV1NumericRules;
    int32?: ListifyV1NumericRules;
    int64?: ListifyV1NumericRules;
    sfixed32?: ListifyV1NumericRules;
    sfixed64?: ListifyV1NumericRules;
    sint32?: ListifyV1NumericRules;
    sint64?: ListifyV1NumericRules;
    uint32?: ListifyV1NumericRules;
    uint64?: ListifyV1NumericRules;
    bool?: ListifyV1BoolRules;
    string?: ListifyV1StringRules;
    enum?: ListifyV1EnumRules;
    timestamp?: ListifyV1TimestampRules; // end oneof "type"
}

export interface ListifyV1NumericRules {
    filterable?: boolean;
    sortable?: boolean;
}

export interface ListifyV1BoolRules {
    filterable?: boolean;
}

export interface ListifyV1StringRules {
    // start oneof "wellKnown"
    openText?: ListifyV1OpenTextRules;
    date?: ListifyV1DateRules;
    foreignKey?: ListifyV1ForeignKeyRules; // end oneof "wellKnown"
}

export interface ListifyV1OpenTextRules {
    searchable?: boolean;
}

export interface ListifyV1DateRules {
    filterable?: boolean;
}

export interface ListifyV1ForeignKeyRules {
    // start oneof "type"
    uniqueString?: ListifyV1UniqueStringRules;
    uuid?: ListifyV1UuidRules; // end oneof "type"
}

export interface ListifyV1UniqueStringRules {
    filterable?: boolean;
}

export interface ListifyV1UuidRules {
    filterable?: boolean;
}

export interface ListifyV1EnumRules {
    filterable?: boolean;
}

export interface ListifyV1TimestampRules {
    filterable?: boolean;
    sortable?: boolean;
}

",
    "path": "generated_from_proto/listify/v1/annotations.ts",
  },
  {
    "content": "export interface ListifyV1PageRequest {
    token?: string;
    limit?: string;
}

export interface ListifyV1PageResponse {
    nextToken?: string;
    finalPage?: boolean;
    totalPageRecords?: string;
    totalRecords?: string;
}

export interface ListifyV1Search {
    query?: string;
}

export interface ListifyV1Sort<T = any, Field extends keyof T | undefined = undefined> {
    field: Field;
    descending?: boolean;
}

export interface ListifyV1Filter<T = any, Field extends keyof T | undefined = undefined> {
    field: Field;
    // start oneof "filterValue"
    value?: string;
    range?: ListifyV1Range; // end oneof "filterValue"
}

export interface ListifyV1FilterClauses {
    clauses?: ListifyV1FilterClause[];
}

export interface ListifyV1FilterClause {
    predicate?: string;
    arguments?: ListifyV1FilterArgument[];
}

export interface ListifyV1FilterArgument {
    // start oneof "kind"
    double?: number;
    fixed32?: number;
    fixed64?: string;
    float?: number;
    int32?: number;
    int64?: string;
    sfixed32?: number;
    sfixed64?: string;
    sint32?: number;
    sint64?: string;
    uint32?: number;
    uint64?: string;
    string?: string; // end oneof "kind"
}

export interface ListifyV1Range {
    // start oneof "type"
    double?: ListifyV1DoubleRange;
    fixed32?: ListifyV1Fixed32Range;
    fixed64?: ListifyV1Fixed64Range;
    float?: ListifyV1FloatRange;
    int32?: ListifyV1Int32Range;
    int64?: ListifyV1Int64Range;
    sfixed32?: ListifyV1SFixed32Range;
    sfixed64?: ListifyV1SFixed64Range;
    sint32?: ListifyV1SInt32Range;
    sint64?: ListifyV1SInt64Range;
    uint32?: ListifyV1UInt32Range;
    uint64?: ListifyV1UInt64Range;
    timestamp?: ListifyV1TimestampRange;
    date?: ListifyV1DateRange; // end oneof "type"
}

export interface ListifyV1DoubleRange {
    min?: number | null;
    max?: number | null;
}

export interface ListifyV1Fixed32Range {
    min?: number | null;
    max?: number | null;
}

export interface ListifyV1Fixed64Range {
    min?: string | null;
    max?: string | null;
}

export interface ListifyV1FloatRange {
    min?: number | null;
    max?: number | null;
}

export interface ListifyV1Int32Range {
    min?: number | null;
    max?: number | null;
}

export interface ListifyV1Int64Range {
    min?: string | null;
    max?: string | null;
}

export interface ListifyV1SFixed32Range {
    min?: number | null;
    max?: number | null;
}

export interface ListifyV1SFixed64Range {
    min?: string | null;
    max?: string | null;
}

export interface ListifyV1SInt32Range {
    min?: number | null;
    max?: number | null;
}

export interface ListifyV1SInt64Range {
    min?: string | null;
    max?: string | null;
}

export interface ListifyV1UInt32Range {
    min?: number | null;
    max?: number | null;
}

export interface ListifyV1UInt64Range {
    min?: string | null;
    max?: string | null;
}

export interface ListifyV1TimestampRange {
    // format: date-time
    min?: string;
    // format: date-time
    max?: string;
}

export interface ListifyV1DateRange {
    min?: string | null;
    max?: string | null;
}

",
    "path": "generated_from_proto/listify/v1/fields.ts",
  },
  {
    "content": "export * from './company_name/services/auth/v1/team';
export * from './company_name/services/auth/v1/user';
export * from './listify/v1/annotations';
export * from './listify/v1/fields';
",
    "path": "generated_from_proto/index.ts",
  },
]
`;
